packages/gatsby/src/schema/infer/inference-metadata.js

```diff
@@ -0,0 +1,407 @@
+/*
+## Incrementally track the structure of nodes with metadata
+
+This metadata can be later utilized for schema inference
+(via building `exampleValue` or directly)
+
+### Usage example:
+
+```javascript
+  const node1 = { id: '1', foo: 25, bar: 'str' }
+  const node2 = { id: '1', foo: 'conflict' }
+
+  let meta = { ignoredFields: new Set(['id']) }
+  meta = addNode(meta, node1)
+  meta = addNode(meta, node2)
+  console.log(meta.fieldMap)
+  // outputs: {
+  //   foo: {
+  //     int: { total: 1, example: 25 },
+  //     string: { total: 1, example: 'conflict' },
+  //   },
+  //   bar: {
+  //     string: { total: 1, example: 'str' },
+  //   },
+  // }
+
+  const example1 = getExampleObject({ meta, typeName, typeConflictReporter })
+  console.log(example1)
+  // outputs { bar: 'str' }
+  // and reports conflicts discovered
+
+  meta = deleteNode(meta, node2)
+  console.log(meta.fieldMap)
+  // outputs: {
+  //   foo: {
+  //     int: { total: 1, example: 25 },
+  //     string: { total: 0, example: 'conflict' },
+  //   },
+  //   bar: { string: { total: 1, example: 'str' } },
+  // }
+
+  const example2 = getExampleObject({ meta, typeName, typeConflictReporter })
+  // outputs: { foo: 25, bar: 'str' }
+```
+
+`addNode`, `deleteNode`, `getExampleObject` are O(N) where N is the number
+of fields in the node object (including nested fields)
+
+### Metadata structure
+
+```javascript
+type TypeMetadata = {
+  ignoredFields?: Set<string>,
+  fieldMap?: { [string]: ValueDescriptor },
+  typeName?: string,
+  dirty?: boolean, // tracks structural changes only
+}
+
+type Count = Number
```

Comment:
is this here a typo with uppercase?

```suggestion
type Count = number
```

Reply:
Yep, thanks for noticing!